sycl/include/syclcompat/memory.hpp

```diff
@@ -651,9 +1016,10 @@ namespace {
 /// \param ptr Point to free.
 /// \returns no return value.
 static inline void free(void *ptr, sycl::queue q = get_default_queue()) {
-  if (ptr) {
-    sycl::free(ptr, q);
-  }
+#ifndef COMPAT_USM_LEVEL_NONE
+  get_device(get_device_id(q.get_device())).queues_wait_and_throw();
+#endif
+  detail::free(ptr, q);
```

Comment:
Yeah well spotted! I have updated the impl to use `detail::free` and tested it where you suggest. I've also removed the `wait_and_throw()` for the USM case, since we've never needed it previously. Feels like a regression, esp since we offer `wait_and_free()`.